[3.0][Store-Upgrade] OOM thrown during schema indices migration

I tried to migrate a 2.3.2 database (size 55GB, approx 1.2 billion nodes and 3 billion relationships) to 3.0.1.

Note: I tried this on my laptop (macbook), the purpose was not to use it extensively, just to be able to retrieve some little informations from it.

During the store upgrade, an OOM was thrown for the schema indices, which is unexpected imo.

**Neo4j Version:** 3.0.1 
**Operating System:** OSX 
**API:** NIL
Default configuration
### Expected behavior

No exception should be thrown, it should just migrate.
### Actual behavior

```
2016-05-12 00:50:06.430+0000 INFO  [o.n.k.i.a.i.IndexPopulationJob] Index population started: [:Langugae(name) [provider: {key=lucene, version=1.0}]]
2016-05-12 01:13:40.470+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Failed to populate index: [:Event(id) [provider: {key=lucene, version=1.0}]] Java heap space
java.lang.OutOfMemoryError: Java heap space
    at org.apache.lucene.index.FreqProxTermsWriterPerField$FreqProxPostingsArray.<init>(FreqProxTermsWriterPerField.java:212)
    at org.apache.lucene.index.FreqProxTermsWriterPerField$FreqProxPostingsArray.newInstance(FreqProxTermsWriterPerField.java:232)
    at org.apache.lucene.index.ParallelPostingsArray.grow(ParallelPostingsArray.java:48)
    at org.apache.lucene.index.TermsHashPerField$PostingsBytesStartArray.grow(TermsHashPerField.java:251)
    at org.apache.lucene.util.BytesRefHash.add(BytesRefHash.java:292)
    at org.apache.lucene.index.TermsHashPerField.add(TermsHashPerField.java:150)
    at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:682)
    at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:365)
    at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:321)
    at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:273)
    at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:413)
    at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1318)
    at org.apache.lucene.index.IndexWriter.addDocuments(IndexWriter.java:1297)
    at org.neo4j.kernel.api.impl.schema.writer.PartitionedIndexWriter.addDocuments(PartitionedIndexWriter.java:63)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.add(LuceneIndexPopulator.java:72)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator.lambda$flush$190(BatchingMultipleIndexPopulator.java:243)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$$Lambda$162/1508954452.run(Unknown Source)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
2016-05-12 01:14:59.154+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Failed to populate index: [:Comment(id) [provider: {key=lucene, version=1.0}]] Java heap space
java.lang.OutOfMemoryError: Java heap space
    at org.apache.lucene.util.BytesRefHash.rehash(BytesRefHash.java:412)
    at org.apache.lucene.util.BytesRefHash.add(BytesRefHash.java:323)
    at org.apache.lucene.index.TermsHashPerField.add(TermsHashPerField.java:150)
    at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:682)
    at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:365)
    at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:321)
    at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:273)
    at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:413)
    at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1318)
    at org.apache.lucene.index.IndexWriter.addDocuments(IndexWriter.java:1297)
    at org.neo4j.kernel.api.impl.schema.writer.PartitionedIndexWriter.addDocuments(PartitionedIndexWriter.java:63)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.add(LuceneIndexPopulator.java:72)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator.lambda$flush$190(BatchingMultipleIndexPopulator.java:243)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$$Lambda$162/1508954452.run(Unknown Source)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
^C~
```

@chrisvest 

Right it finally failed completely.

heap-dump : https://s3-eu-west-1.amazonaws.com/graphaware/assets/heap-dump.hprof

`logs/debug.log` :

```
2016-05-12 11:23:15.982+0000 INFO  [o.n.k.i.a.i.IndexPopulationJob] Forcefully shutting down executor.
BatchingMultipleIndexPopulator{activeTasks=4, executor=java.util.concurrent.ThreadPoolExecutor@64cc4088[Running, pool size = 3, active threads = 3, queued tasks = 1, completed tasks = 3314], batchedUpdates = [org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@48d6fd54 - 1616 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@7b96b334 - 3500 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@6aad84ee - 677 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@15ea2fcf - 1616 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@19542291 - 2508 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@603c5299 - 274 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@181ed198 - 2003 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@29517119 - 2985 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@24ec959c - 274 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@2475f4c7 - 8329 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@6defcdda - 9601 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@49c37ad5 - 4628 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@512de9 - 7313 updates, org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingIndexPopulation@11a98917 - 2003 updates], queuedUpdates = 0}
2016-05-12 11:23:15.995+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Failed to populate index: [:Repository(id) [provider: {key=lucene, version=1.0}]] Java heap space
java.lang.OutOfMemoryError: Java heap space
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure$DocWithId.setId(LuceneDocumentStructure.java:283)
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure$DocWithId.access$100(LuceneDocumentStructure.java:261)
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure.reuseDocument(LuceneDocumentStructure.java:89)
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure.documentRepresentingProperty(LuceneDocumentStructure.java:95)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.updateAsDocument(LuceneIndexPopulator.java:99)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator$$Lambda$158/1618342786.apply(Unknown Source)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1351)
    at java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$222(StreamSpliterators.java:294)
    at java.util.stream.StreamSpliterators$WrappingSpliterator$$Lambda$163/1612794450.getAsBoolean(Unknown Source)
    at java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
    at java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:169)
    at java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:300)
    at java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
    at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:259)
    at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:413)
    at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1318)
    at org.apache.lucene.index.IndexWriter.addDocuments(IndexWriter.java:1297)
    at org.neo4j.kernel.api.impl.schema.writer.PartitionedIndexWriter.addDocuments(PartitionedIndexWriter.java:63)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.add(LuceneIndexPopulator.java:72)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator.lambda$flush$190(BatchingMultipleIndexPopulator.java:243)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$$Lambda$157/1540967816.run(Unknown Source)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
2016-05-12 11:23:15.999+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Failed to populate index: [:Repository(full_name) [provider: {key=lucene, version=1.0}]] Java heap space
java.lang.OutOfMemoryError: Java heap space
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure$DocWithId.setId(LuceneDocumentStructure.java:283)
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure$DocWithId.access$100(LuceneDocumentStructure.java:261)
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure.reuseDocument(LuceneDocumentStructure.java:89)
    at org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure.documentRepresentingProperty(LuceneDocumentStructure.java:95)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.updateAsDocument(LuceneIndexPopulator.java:99)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator$$Lambda$158/1618342786.apply(Unknown Source)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1351)
    at java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$222(StreamSpliterators.java:294)
    at java.util.stream.StreamSpliterators$WrappingSpliterator$$Lambda$163/1612794450.getAsBoolean(Unknown Source)
    at java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
    at java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:169)
    at java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:300)
    at java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
    at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:259)
    at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:413)
    at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1318)
    at org.apache.lucene.index.IndexWriter.addDocuments(IndexWriter.java:1297)
    at org.neo4j.kernel.api.impl.schema.writer.PartitionedIndexWriter.addDocuments(PartitionedIndexWriter.java:63)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.add(LuceneIndexPopulator.java:72)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator.lambda$flush$190(BatchingMultipleIndexPopulator.java:243)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$$Lambda$157/1540967816.run(Unknown Source)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
2016-05-12 11:23:25.688+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Unable to close failed populator for index: [:Repository(id) [provider: {key=lucene, version=1.0}]]
java.nio.channels.ClosedByInterruptException
    at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
    at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:216)
    at org.neo4j.io.fs.StoreFileChannel.write(StoreFileChannel.java:153)
    at org.neo4j.kernel.api.impl.index.storage.FailureStorage.storeIndexFailure(FailureStorage.java:119)
    at org.neo4j.kernel.api.impl.index.storage.PartitionedIndexStorage.storeIndexFailure(PartitionedIndexStorage.java:114)
    at org.neo4j.kernel.api.impl.schema.LuceneSchemaIndex.markAsFailed(LuceneSchemaIndex.java:197)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.markAsFailed(LuceneIndexPopulator.java:94)
    at org.neo4j.kernel.impl.api.index.MultipleIndexPopulator.fail(MultipleIndexPopulator.java:220)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator.lambda$flush$190(BatchingMultipleIndexPopulator.java:247)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
2016-05-12 11:23:25.689+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Unable to close failed populator for index: [:Repository(full_name) [provider: {key=lucene, version=1.0}]]
java.nio.channels.ClosedByInterruptException
    at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
    at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:216)
    at org.neo4j.io.fs.StoreFileChannel.write(StoreFileChannel.java:153)
    at org.neo4j.kernel.api.impl.index.storage.FailureStorage.storeIndexFailure(FailureStorage.java:119)
    at org.neo4j.kernel.api.impl.index.storage.PartitionedIndexStorage.storeIndexFailure(PartitionedIndexStorage.java:114)
    at org.neo4j.kernel.api.impl.schema.LuceneSchemaIndex.markAsFailed(LuceneSchemaIndex.java:197)
    at org.neo4j.kernel.api.impl.schema.populator.LuceneIndexPopulator.markAsFailed(LuceneIndexPopulator.java:94)
    at org.neo4j.kernel.impl.api.index.MultipleIndexPopulator.fail(MultipleIndexPopulator.java:220)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator.lambda$flush$190(BatchingMultipleIndexPopulator.java:247)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
2016-05-12 11:23:25.693+0000 ERROR [o.n.k.i.a.i.IndexPopulationJob] Failed to populate index: [:Branch(id) [provider: {key=lucene, version=1.0}]] Java heap space
java.lang.OutOfMemoryError: Java heap space
    at org.neo4j.kernel.impl.store.InlineNodeLabels.parseInlined(InlineNodeLabels.java:149)
    at org.neo4j.kernel.impl.store.InlineNodeLabels.get(InlineNodeLabels.java:57)
    at org.neo4j.kernel.impl.store.InlineNodeLabels.get(InlineNodeLabels.java:52)
    at org.neo4j.kernel.impl.transaction.state.NeoStoreIndexStoreView$StoreViewNodeStoreScan.process(NeoStoreIndexStoreView.java:268)
    at org.neo4j.kernel.impl.transaction.state.NeoStoreIndexStoreView$NodeStoreScan.run(NeoStoreIndexStoreView.java:217)
    at org.neo4j.kernel.impl.api.index.BatchingMultipleIndexPopulator$BatchingStoreScan.run(BatchingMultipleIndexPopulator.java:384)
    at org.neo4j.kernel.impl.api.index.IndexPopulationJob.indexAllNodes(IndexPopulationJob.java:138)
    at org.neo4j.kernel.impl.api.index.IndexPopulationJob.run(IndexPopulationJob.java:110)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:104)
2016-05-12 11:23:25.704+0000 WARN  [o.n.k.NeoStoreDataSource] Exception occurred while starting the datasource. Attempting to close things down.
2016-05-12 11:23:25.812+0000 INFO  [o.n.k.i.DiagnosticsManager] --- STOPPING diagnostics START ---
2016-05-12 11:23:25.813+0000 INFO  [o.n.k.i.DiagnosticsManager] --- STOPPING diagnostics END ---
2016-05-12 11:23:25.814+0000 INFO  [o.n.k.i.e.EnterpriseFacadeFactory] Shutdown started
```

`logs/neo4j.log` : 

```
Dumping heap to heap-dump.hprof ...
Heap dump file created [2690509565 bytes in 28.177 secs]
Exception in thread "Lucene Merge Thread #31" org.apache.lucene.index.MergePolicy$MergeException: java.lang.OutOfMemoryError: Java heap space
    at org.apache.lucene.index.ConcurrentMergeScheduler.handleMergeException(ConcurrentMergeScheduler.java:668)
    at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:648)
Caused by: java.lang.OutOfMemoryError: Java heap space
    at org.apache.lucene.codecs.blocktreeords.FSTOrdsOutputs.newOutput(FSTOrdsOutputs.java:225)
    at org.apache.lucene.codecs.blocktreeords.FSTOrdsOutputs.subtract(FSTOrdsOutputs.java:145)
    at org.apache.lucene.codecs.blocktreeords.FSTOrdsOutputs.subtract(FSTOrdsOutputs.java:32)
    at org.apache.lucene.util.fst.Builder.add(Builder.java:455)
    at org.apache.lucene.codecs.blocktreeords.OrdsBlockTreeTermsWriter$PendingBlock.append(OrdsBlockTreeTermsWriter.java:422)
    at org.apache.lucene.codecs.blocktreeords.OrdsBlockTreeTermsWriter$PendingBlock.compileIndex(OrdsBlockTreeTermsWriter.java:387)
    at org.apache.lucene.codecs.blocktreeords.OrdsBlockTreeTermsWriter$TermsWriter.writeBlocks(OrdsBlockTreeTermsWriter.java:555)
    at org.apache.lucene.codecs.blocktreeords.OrdsBlockTreeTermsWriter$TermsWriter.finish(OrdsBlockTreeTermsWriter.java:848)
    at org.apache.lucene.codecs.blocktreeords.OrdsBlockTreeTermsWriter.write(OrdsBlockTreeTermsWriter.java:240)
    at org.apache.lucene.codecs.perfield.PerFieldPostingsFormat$FieldsWriter.write(PerFieldPostingsFormat.java:198)
    at org.apache.lucene.codecs.FieldsConsumer.merge(FieldsConsumer.java:105)
    at org.apache.lucene.index.SegmentMerger.mergeTerms(SegmentMerger.java:193)
    at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:95)
    at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:4075)
    at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:3655)
    at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:588)
    at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:626)
2016-05-12 11:23:26.417+0000 ERROR Failed to start Neo4j: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@7ea3276e' was successfully initialized, but failed to start. Please see attached cause exception. Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@7ea3276e' was successfully initialized, but failed to start. Please see attached cause exception.
org.neo4j.server.ServerStartupException: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@7ea3276e' was successfully initialized, but failed to start. Please see attached cause exception.
    at org.neo4j.server.exception.ServerStartupErrors.translateToServerStartupError(ServerStartupErrors.java:68)
    at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:217)
    at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:81)
    at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:60)
    at org.neo4j.server.enterprise.EnterpriseEntryPoint.main(EnterpriseEntryPoint.java:32)
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.server.database.LifecycleManagingDatabase@7ea3276e' was successfully initialized, but failed to start. Please see attached cause exception.
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:444)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:189)
    ... 3 more
Caused by: java.lang.RuntimeException: Error starting org.neo4j.kernel.impl.enterprise.EnterpriseFacadeFactory, /Users/ikwattro/dev/_graphs/github3/data/databases/graph.db
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:144)
    at org.neo4j.kernel.impl.enterprise.EnterpriseFacadeFactory.newFacade(EnterpriseFacadeFactory.java:42)
    at org.neo4j.graphdb.EnterpriseGraphDatabase.<init>(EnterpriseGraphDatabase.java:57)
    at org.neo4j.server.enterprise.EnterpriseNeoServer.lambda$static$1(EnterpriseNeoServer.java:85)
    at org.neo4j.server.database.LifecycleManagingDatabase.start(LifecycleManagingDatabase.java:89)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:434)
    ... 5 more
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.impl.storageengine.impl.recordstorage.RecordStorageEngine@3d57718a' was successfully initialized, but failed to start. Please see attached cause exception.
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:444)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:508)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:434)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:99)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:434)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:140)
    ... 10 more
Caused by: java.lang.IllegalStateException: Index entered FAILED state while recovery waited for it to be fully populated
    at org.neo4j.kernel.impl.api.index.IndexingService.awaitOnline(IndexingService.java:339)
    at org.neo4j.kernel.impl.api.index.IndexingService.start(IndexingService.java:319)
    at org.neo4j.kernel.impl.storageengine.impl.recordstorage.RecordStorageEngine.start(RecordStorageEngine.java:413)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:434)
    ... 18 more
```

--- a/community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
+++ b/community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
@@ -303,10 +303,17 @@ public abstract class GraphDatabaseSettings
     public static final Setting<Boolean> index_background_sampling_enabled =
             setting("dbms.index_sampling.background_enabled", BOOLEAN, TRUE );
 
-    @Description("Size of buffer used by index sampling")
-    public static final Setting<Long> index_sampling_buffer_size =
-            setting("dbms.index_sampling.buffer_size", BYTES, "8m",
+    @Description("Size of buffer used by index sampling. " +
+                 "This configuration setting is no longet applicable as from Neo4j 3.0.3." +
+                 "Please use dbms.index_sampling.sample_size_limit instead.")
+    @Deprecated
+    public static final Setting<Long> index_sampling_buffer_size = setting("dbms.index_sampling.buffer_size", BYTES, "64m",
                     min( /* 1m */ 1048576L ), max( (long) Integer.MAX_VALUE ) );
+
+    @Description("Index sampling chunk size limit")
+    public static final Setting<Integer> index_sample_size_limit = setting("dbms.index_sampling.sample_size_limit",
+            INTEGER, String.valueOf( ByteUnit.mebiBytes( 8 ) ), min( (int) ByteUnit.mebiBytes( 1 ) ),
+            max( Integer.MAX_VALUE ) );
 
     @Description("Percentage of index updates of total index size required before sampling of a given index is triggered")
     public static final Setting<Integer> index_sampling_update_percentage =

--- a/community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
+++ b/community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
@@ -304,7 +304,7 @@ public abstract class GraphDatabaseSettings
             setting("dbms.index_sampling.background_enabled", BOOLEAN, TRUE );
 
     @Description("Size of buffer used by index sampling. " +
-                 "This configuration setting is no longet applicable as from Neo4j 3.0.3." +
+                 "This configuration setting is no longer applicable as from Neo4j 3.0.3." +
                  "Please use dbms.index_sampling.sample_size_limit instead.")
     @Deprecated
     public static final Setting<Long> index_sampling_buffer_size = setting("dbms.index_sampling.buffer_size", BYTES, "64m",

--- a/community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
+++ b/community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
@@ -305,7 +305,7 @@ public abstract class GraphDatabaseSettings
 
     @Description("Size of buffer used by index sampling")
     public static final Setting<Long> index_sampling_buffer_size =
-            setting("dbms.index_sampling.buffer_size", BYTES, "64m",
+            setting("dbms.index_sampling.buffer_size", BYTES, "8m",
                     min( /* 1m */ 1048576L ), max( (long) Integer.MAX_VALUE ) );
 
     @Description("Percentage of index updates of total index size required before sampling of a given index is triggered")

--- a/community/kernel/src/main/java/org/neo4j/kernel/configuration/GraphDatabaseConfigurationMigrator.java
+++ b/community/kernel/src/main/java/org/neo4j/kernel/configuration/GraphDatabaseConfigurationMigrator.java
@@ -19,6 +19,39 @@
  */
 package org.neo4j.kernel.configuration;
 
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Map;
+
+import org.neo4j.graphdb.factory.GraphDatabaseSettings;
+import org.neo4j.io.ByteUnit;
+
+/**
+ * Migrations of old graph database settings.
+ */
 public class GraphDatabaseConfigurationMigrator extends BaseConfigurationMigrator
 {
+    public GraphDatabaseConfigurationMigrator()
+    {
+        registerMigrations();
+    }
+
+    private void registerMigrations()
+    {
+        add( new SpecificPropertyMigration( "dbms.index_sampling.buffer_size",
+                "dbms.index_sampling.buffer_size has been replaced with dbms.index_sampling.sample_size_limit" )
+        {
+            @Override
+            public void setValueWithOldSetting( String value, Map<String,String> rawConfiguration )
+            {
+                if ( StringUtils.isNotEmpty( value ) )
+                {
+                    String oldSettingDefaultValue = GraphDatabaseSettings.index_sampling_buffer_size.getDefaultValue();
+                    Long newValue = oldSettingDefaultValue.equals( value ) ? ByteUnit.mebiBytes( 8 )
+                                                                           : Settings.BYTES.apply( value );
+                    rawConfiguration.put( "dbms.index_sampling.sample_size_limit", String.valueOf( newValue ) );
+                }
+            }
+        } );
+    }
 }

--- a/community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingConfig.java
+++ b/community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingConfig.java
@@ -30,7 +30,7 @@ public class IndexSamplingConfig
 
     public IndexSamplingConfig( Config config )
     {
-        this.sampleSizeLimit = config.get( GraphDatabaseSettings.index_sampling_buffer_size ).intValue();
+        this.sampleSizeLimit = config.get( GraphDatabaseSettings.index_sample_size_limit );
         this.updateRatio = ((double) config.get( GraphDatabaseSettings.index_sampling_update_percentage )) / 100.0d;
         this.backgroundSampling = config.get( GraphDatabaseSettings.index_background_sampling_enabled );
     }

--- a/community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingConfig.java
+++ b/community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingConfig.java
@@ -24,20 +24,20 @@ import org.neo4j.kernel.configuration.Config;
 
 public class IndexSamplingConfig
 {
-    private final int bufferSize;
+    private final int sampleSizeLimit;
     private final double updateRatio;
     private final boolean backgroundSampling;
 
     public IndexSamplingConfig( Config config )
     {
-        this.bufferSize = config.get( GraphDatabaseSettings.index_sampling_buffer_size ).intValue();
+        this.sampleSizeLimit = config.get( GraphDatabaseSettings.index_sampling_buffer_size ).intValue();
         this.updateRatio = ((double) config.get( GraphDatabaseSettings.index_sampling_update_percentage )) / 100.0d;
         this.backgroundSampling = config.get( GraphDatabaseSettings.index_background_sampling_enabled );
     }
 
-    public int bufferSize()
+    public int sampleSizeLimit()
     {
-        return bufferSize;
+        return sampleSizeLimit;
     }
 
     public double updateRatio()
@@ -71,14 +71,14 @@ public class IndexSamplingConfig
         IndexSamplingConfig that = (IndexSamplingConfig) o;
 
         return backgroundSampling == that.backgroundSampling &&
-               bufferSize == that.bufferSize &&
+               sampleSizeLimit == that.sampleSizeLimit &&
                Double.compare( that.updateRatio, updateRatio ) == 0;
     }
 
     @Override
     public int hashCode()
     {
-        int result = bufferSize;
+        int result = sampleSizeLimit;
         long temp = Double.doubleToLongBits( updateRatio );
         result = 31 * result + (int) (temp ^ (temp >>> 32));
         result = 31 * result + (backgroundSampling ? 1 : 0);

--- a/community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/NonUniqueIndexSampler.java
+++ b/community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/NonUniqueIndexSampler.java
@@ -24,9 +24,7 @@ import org.neo4j.storageengine.api.schema.IndexSample;
 
 public class NonUniqueIndexSampler
 {
-    private static final int INITIAL_SIZE = 1 << 16;
-
-    private final int bufferSizeLimit;
+    private final int sampleSizeLimit;
     private final MultiSet<String> values;
 
     private int sampledSteps = 0;
@@ -35,12 +33,12 @@ public class NonUniqueIndexSampler
 
     private long accumulatedUniqueValues = 0;
     private long accumulatedSampledSize = 0;
-    private long bufferSize = 0;
+    private long sampleSize = 0;
 
-    public NonUniqueIndexSampler( int bufferSizeLimit )
+    public NonUniqueIndexSampler( int sampleSizeLimit )
     {
-        this.bufferSizeLimit = bufferSizeLimit;
-        this.values = new MultiSet<>( INITIAL_SIZE );
+        this.values = new MultiSet<>( calculateInitialSetSize( sampleSizeLimit ) );
+        this.sampleSizeLimit = sampleSizeLimit;
     }
 
     public void include( String value )
@@ -51,14 +49,14 @@ public class NonUniqueIndexSampler
     public void include( String value, long increment )
     {
         assert increment > 0;
-        if ( bufferSize >= bufferSizeLimit )
+        if ( sampleSize >= sampleSizeLimit )
         {
             nextStep();
         }
 
         if ( values.increment( value, increment ) == increment )
         {
-            bufferSize += value.length();
+            sampleSize += value.length();
         }
     }
 
@@ -72,7 +70,7 @@ public class NonUniqueIndexSampler
         assert decrement > 0;
         if ( values.increment( value, -decrement ) == 0 )
         {
-            bufferSize -= value.length();
+            sampleSize -= value.length();
         }
     }
 
@@ -98,9 +96,24 @@ public class NonUniqueIndexSampler
     {
         accumulatedUniqueValues += values.uniqueSize();
         accumulatedSampledSize += values.size();
-        bufferSize = 0;
+        sampleSize = 0;
 
         sampledSteps++;
         values.clear();
+    }
+
+    /**
+     * Evaluate initial set size that evaluate initial set as log2(sampleSizeLimit) / 2 based on provided sample size
+     * limit.
+     * Minimum possible size is 1 << 10.
+     * Maximum possible size is 1 << 16.
+     *
+     * @param sampleSizeLimit specified sample size limit
+     * @return initial set size
+     */
+    private int calculateInitialSetSize( int sampleSizeLimit )
+    {
+        int basedOnSampleSize = Math.max( 10, (int) (Math.log( sampleSizeLimit ) / Math.log( 2 )) / 2 );
+        return (1 << Math.min( 16, basedOnSampleSize ));
     }
 }

--- a/community/kernel/src/test/java/org/neo4j/kernel/configuration/TestGraphDatabaseConfigurationMigrator.java
+++ b/community/kernel/src/test/java/org/neo4j/kernel/configuration/TestGraphDatabaseConfigurationMigrator.java
@@ -21,11 +21,13 @@ package org.neo4j.kernel.configuration;
 
 import org.junit.Test;
 
+import java.util.Map;
+
 import org.neo4j.logging.NullLog;
 
 import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertThat;
-
 import static org.neo4j.helpers.collection.MapUtil.stringMap;
 
 /**
@@ -36,7 +38,25 @@ public class TestGraphDatabaseConfigurationMigrator
     @Test
     public void testNoMigration()
     {
-        ConfigurationMigrator migrator = new GraphDatabaseConfigurationMigrator(  );
+        ConfigurationMigrator migrator = new GraphDatabaseConfigurationMigrator();
         assertThat( migrator.apply( stringMap( "foo", "bar" ), NullLog.getInstance() ), equalTo( stringMap( "foo", "bar" ) ) );
+    }
+
+    @Test
+    public void migrateIndexSamplingBufferSizeIfPresent()
+    {
+        ConfigurationMigrator migrator = new GraphDatabaseConfigurationMigrator();
+        Map<String,String> resultConfig = migrator.apply( stringMap( "dbms.index_sampling.buffer_size", "64m" ), NullLog.getInstance() );
+        assertEquals( "Old property should be migrated to new one with correct value",
+                resultConfig, stringMap( "dbms.index_sampling.sample_size_limit", "8388608" ));
+    }
+
+    @Test
+    public void skipMigrationOfIndexSamplingBufferSizeIfNotPresent()
+    {
+        ConfigurationMigrator migrator = new GraphDatabaseConfigurationMigrator();
+        Map<String,String> resultConfig = migrator.apply( stringMap( "dbms.index_sampling.sample_size_limit", "8388600" ), NullLog.getInstance() );
+        assertEquals( "Nothing to migrate should be the same",
+                resultConfig, stringMap( "dbms.index_sampling.sample_size_limit", "8388600" ));
     }
 }

--- a/community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/populator/NonUniqueLuceneIndexPopulator.java
+++ b/community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/populator/NonUniqueLuceneIndexPopulator.java
@@ -42,7 +42,7 @@ public class NonUniqueLuceneIndexPopulator extends LuceneIndexPopulator
     public NonUniqueLuceneIndexPopulator( LuceneSchemaIndex luceneIndex, IndexSamplingConfig samplingConfig )
     {
         super( luceneIndex );
-        this.sampler = new NonUniqueIndexSampler( samplingConfig.bufferSize() );
+        this.sampler = new NonUniqueIndexSampler( samplingConfig.sampleSizeLimit() );
     }
 
     @Override

--- a/community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/sampler/NonUniqueLuceneIndexSampler.java
+++ b/community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/sampler/NonUniqueLuceneIndexSampler.java
@@ -58,7 +58,7 @@ public class NonUniqueLuceneIndexSampler extends LuceneIndexSampler
     @Override
     protected IndexSample performSampling() throws IndexNotFoundKernelException
     {
-        NonUniqueIndexSampler sampler = new NonUniqueIndexSampler( indexSamplingConfig.bufferSize() );
+        NonUniqueIndexSampler sampler = new NonUniqueIndexSampler( indexSamplingConfig.sampleSizeLimit() );
         IndexReader indexReader = indexSearcher.getIndexReader();
         for ( LeafReaderContext readerContext : indexReader.leaves() )
         {

